Src/main.c

```diff
@@ -106,15 +112,25 @@ int main(void)
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
+
+  const uint8_t TIMER_LOW = 1;
+  const uint8_t TIMER_HIGH = 2;
+
   while (1)
   {
-  /* USER CODE END WHILE */
 
+  /* USER CODE END WHILE */
+    
   /* USER CODE BEGIN 3 */
+    uint16_t adcValue = HAL_ADC_GetState(&hadc);
+    float advValPercentage = adcValue/4096.0; // Converting the converted digital value into a percentage
+
+    float compareVal = (advValPercentage * (TIMER_HIGH - TIMER_LOW) + TIMER_LOW) * htim16.Init.Period / (float) TIMER_HIGH; // Scales the val percentage to a value between 2000 and 4000 ticks, in line with the duty cycle and set prescaler
```

Comment:
math isn't correct. I'd recommend trying this in an online compiler or something to test some examples.